File: app/lib/reference_code.rb
# Generates and validates the rider-facing booking lookup code. See
# kos/decisions/reference-code-format.md: 6 random chars from an unambiguous 31-char alphabet
# (no 0/O, 1/I/L) plus a 1-char Luhn-mod-N checksum, so staff hand-typing this at the boarding
# counter catch a single mistyped character immediately.
#
# `Booking#reference_code` stores the raw 7-char code (no dashes); `.format` renders the
# display grouping (XXX-XXX-C) and lookups strip the dashes again before calling `.valid?`.
module ReferenceCode
  ALPHABET = "23456789ABCDEFGHJKMNPQRSTUVWXYZ".freeze # N = 31, no 0/1/I/L/O

  BODY_LENGTH = 6

  def self.generate
    body = Array.new(BODY_LENGTH) { ALPHABET[SecureRandom.random_number(ALPHABET.length)] }.join
    body + checksum(body)
  end

  def self.checksum(code)
    n = ALPHABET.length
    factor = 2
    sum = 0
    code.reverse.each_char do |char|
      addend = factor * ALPHABET.index(char)
      factor = factor == 2 ? 1 : 2
      addend = (addend / n) + (addend % n)
      sum += addend
    end
    remainder = sum % n
    ALPHABET[(n - remainder) % n]
  end

  def self.valid?(full_code)
    return false if full_code.blank? || full_code.length != BODY_LENGTH + 1

    body, check = full_code[0..-2], full_code[-1]
    return false unless (body + check).chars.all? { |c| ALPHABET.include?(c) }

    checksum(body) == check
  end

  def self.format(full_code)
    body = full_code[0..-2]
    "#{body[0..2]}-#{body[3..5]}-#{full_code[-1]}"
  end
end

File: frontend/src/utils/referenceCode.ts
// Client-side mirror of app/lib/reference_code.rb's Luhn-mod-N checksum — lets Booking Lookup
// catch a mistyped character before it ever reaches the server (see
// kos/decisions/reference-code-format.md and kos/decisions/ux/mockups/booking-lookup.html's
// "Local — checksum fails" state). Keep this in lockstep with the Ruby module if either changes.

const ALPHABET = '23456789ABCDEFGHJKMNPQRSTUVWXYZ' // N = 31, no 0/1/I/L/O
const BODY_LENGTH = 6

// Strips dashes/spaces and uppercases — mirrors BookingLookup's own normalization server-side, so
// "4xk-7qm-9" and "4XK7QM9" both validate/submit the same as "4XK-7QM-9".
export function normalizeReferenceCode(input: string): string {
  return input.replace(/[-\s]/g, '').toUpperCase()
}

function checksum(body: string): string {
  const n = ALPHABET.length
  let factor = 2
  let sum = 0
  for (const char of [...body].reverse()) {
    let addend = factor * ALPHABET.indexOf(char)
    factor = factor === 2 ? 1 : 2
    addend = Math.floor(addend / n) + (addend % n)
    sum += addend
  }
  const remainder = sum % n
  return ALPHABET[(n - remainder) % n]
}

// Expects an already-normalized code (see normalizeReferenceCode). False for anything the wrong
// length or containing a character outside the alphabet, not just a checksum mismatch.
export function isValidReferenceCode(code: string): boolean {
  if (code.length !== BODY_LENGTH + 1) return false

  const body = code.slice(0, -1)
  const check = code.slice(-1)
  if (![...body, check].every((c) => ALPHABET.includes(c))) return false

  return checksum(body) === check
}
